File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
    void findCombination(int index, vector<int>& arr, vector<int>& ds,int target,vector<vector<int>>& ans)
    {
        if(index == arr.size())
        {
            if(target ==0) {
                ans.push_back(ds);
            }
             return;
        }
        
        //Pick the element 
        if(arr[index] <= target)
        {
            ds.push_back(arr[index]);
            findCombination(index, arr, ds, target-arr[index], ans);
            ds.pop_back();
        }
        
        findCombination(index+1, arr, ds, target, ans);
    }
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> ds;
        findCombination(0,candidates,ds,target,ans);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public void helper(int idx, int target, int[]candidates,List<List<Integer>> ans, List<Integer>curr_ds){
        if(idx == candidates.length){
            if(target == 0){
                ans.add(new ArrayList<>(curr_ds));
            }
              return ;
        }

        if(candidates[idx] <= target){
            curr_ds.add(candidates[idx]);
            helper(idx, target-candidates[idx], candidates, ans, curr_ds);
            curr_ds.remove(curr_ds.size()-1);
        }

        helper(idx+1, target, candidates, ans, curr_ds);
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> curr_ds = new ArrayList<>();

        helper(0,target,candidates,ans,curr_ds);

        return ans;
    }
}